scripts/builtin/residencyMatchMain.dml

```diff
@@ -0,0 +1,194 @@
+#
+# Capacity.mtx
+# 1.0
+# 1.0
+# 1.0
+#
+# ResidencyMatch.mtx
+# 0.0,0.0,3.0
+# 1.0,0.0,0.0
+# 0.0,2.0,0.0
+#
+# HospitalMatch.mtx
+# 0.0,1.0,0.0
+# 0.0,0.0,2.0
+# 1.0,0.0,0.0
+#
+# Resident 1 has matched with Hospital 3 (since [1,3] is non-zero) at a preference level of 3.0.
+# Resident 2 has matched with Hospital 1 (since [2,1] is non-zero) at a preference level of 1.0.
+# Resident 3 has matched with Hospital 2 (since [3,2] is non-zero) at a preference level of 2.0.
+# --------------------------------------------------------------------------------------------
+m_residencyMatchMain = function(Matrix[Double] R, Matrix[Double] H,Matrix[Double] C )
+  return (Matrix[Double] ResidencyMatch)
+{
+#TODO PLEASE PAY ATTENTION
+# in this step we consider that  Residents Matrix is ORDERED   !!!!
+# in this step we consider that  Hospital  Matrix is UNORDERED !!!!
+
+print("STARTING Resident Hospital Match");
+#TODO set a finite number of maximum iterations so that the execution termiates after maximum iterations.
+
+print("\n")
+print("STARTING RESIDENCY MATCH ALGORITHM");
+print("READING R  as residents AND H as Hospitals and also C as capacity...");
+
+m = nrow(R)
+n = ncol (R)
+Capacityrows = nrow(C)
+#######################################################################################################
+Capacity = matrix(0.0, rows=Capacityrows , cols=1)
+Capacity = C;
+max_position = colMaxs(Capacity)
+#######################################################################################################
+# we can consider number of choices for every resident #
+#######################################################################################################
+ResidencyMatch = matrix(0.0, rows=m, cols=n)
+HospitalMatch = matrix(0.0, rows=n, cols=m)
+Result_matrix = matrix(0.0, rows=nrow(R), cols=ncol(R))
+########################################################################################################
+if(nrow(Capacity) != nrow(H)) {
+  print("ERROR: Wrong Input !Capacity indexes is not match with the Number of hospitals ")
+  #it means that some hospitals' capacity is not defined.
+}# end of if
+
+StarM = matrix(1.0, rows=m, cols=1)  ### for checking while
+
+HIndex =matrix(1.0, rows=m, cols=1)
+proposer_pointers = matrix(1.0, rows=m, cols=1)
+prev_Residents_vector = matrix(1.0, rows=n, cols=1)
+prevIndex_Residents_vector = matrix(1.0, rows=n, cols=1)
+
+prev_Residents_vector = rowMins(HospitalMatch)
+prevIndex_Residents_vector =  rowIndexMin(HospitalMatch)
+
+while(sum(StarM) > 0) {
+
+  for(i in 1:m) {
+
+    while (as.scalar (StarM[i]) == 1)  {
+      SecondIndex = as.scalar (proposer_pointers[i])
+      HIndex[i] = as.scalar (R[i,SecondIndex])
+      prev_Residents_vector = rowMaxs(HospitalMatch) # we consider if the preference value is 1 it means it is the first selection of that hospital, the minimum value means most preference.
```

Comment:
Please set a line-width/margin to 80 - 100 characters in your editor to break long lines.